### ISSUE TITLE
[2017 ICLR] A Baseline for Detecting Misclassified and Out-of-Distribution Examples in Neural Networks

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#59]
모델이 inference 때 misclassifed / out-of-distribution input을 받더라도 maximum softmax probability값이 크기에 이를 confidence로 보기엔 어려움이 있음 
(softmax는 indicator function의 smooth approximation이기에)

그럼에도 불구하고 maximum softmax probability를 이용하여 correctly classified examples인지, misclassified / out-of-distribution인지 detect할 수 있음을 보여준 논문

즉 softmax classifier probability가 confidence를 직접적으로 측정하기에 유용하지 않지만, model의 confidence를 측정하는데 사용될 수 있음을 보여줌

### 1.1. Problem Formulation

논문에서의 주 관심사는 2가지임

1. Error and success prediction
    - can we predict whether a trained classifier will make an error on a particular held-out test examples?
    - can we predict if it will correctly classify said example?
2. In-distribution and out-of-distribution detection
    - can we predict whether a test example is from a different distribution from the training data?
    - can we predict if it is from within the same distribution?

### 1.2. Measure
![image](https://user-images.githubusercontent.com/49643709/160864373-42d5a7ef-fb0b-48c4-b038-f71040f8e3c0.png)

> - AUROC (Area Under the Receiver Operating Characteristic curve) : 50 ~ 100 %
> - AUPR (Area Under the Precision-Recall curve) : precision ~ 100 %
> - ROC curve : TPR (True Positive Rate), FPR (False Positive Rate)
> - PR curve : precision, recall

예를 들어 2-class classification 문제라고 한다면 모델의 output은 positive class, negative class임

Accuracy를 measure로 설정하게 된다면 positive가 negative보다 얼마나 더 likely한지 상관 없이 positive만 선택하게 됨
(vice versa, argmax값을 통해 정해지지 그 정도에 대한 생각은 안하기에)

이러한 이슈를 해결하기 위해 score threshold를 통해 positive인지 아닌지 등을 결정

Threshold-independent performance evaluation으로 AUROC와 AUPR을 사용
(AUROC is not ideal when the positive class and negative class have greatly differing base rates)

AUPR 또한 positive class의 base rate에 영향을 많이 받기에 success/normal classes를 positive보는 경우, error/abnormal을 positive로 보는 경우로 따로 측정
(error/abnormal를 positive로 보는 경우, score인 probability 값에 -1을 곱해서 계산)

 
### 2.1. Baseline
![image](https://user-images.githubusercontent.com/49643709/161045893-d079afb1-e431-40b8-b3b0-6e741c18bfd4.png)

> - 'Succ' : correctly classified test set examples as positive
> - 'Err' : incorrectly classified test set examples as positive
> - 'In' : in-distribution, correctly classified test set examples as positive
> - 'Out' : out-of-distribution examples as positive
> - 'Pred Prob Wrong (mean)' : mean predicted class probability of wrongly classified examples
> - 'Pred Prob (mean)' : mean predicted class probability of out-of-distribution examples

Softmax distribution의 maximum/predicted class probability 값을 이용하여 example이 erroneously classified인지 out-of-distribution인지 detect하는 것을 baseline으로 잡음

다음과 같은 과정을 통해 AUPR, AUROC를 구함

1. Test set examples에 대하여 correctly classified / incorrectly classified로 나눔 (2 class)
2. 각 sample에 대해 predicted class에 대한 softmax probability 값을 구함 (maximum softmax probability)
3. 2개의 그룹으로부터 AUROC, AUPR을 얻음

### 2.2. Computer Vision
![image](https://user-images.githubusercontent.com/49643709/160889232-7db73277-3574-44f5-924d-a8279ba2d81b.png)
![image](https://user-images.githubusercontent.com/49643709/160889921-8b7bae11-5697-493b-87ca-37836fd09c0c.png)

> - CIFAR-10bw : black and white rescaled CIFAR-10 images
> - Gaussian : random normal noise
> - Uniform : random uniform noise

Table 1를 보면 correctly classified examples와 incorrectly classified examples는 충분히 다르며 구분할 수 있음

Table 2를 보면 in-distribution examples와 out-of-distribution examples는 충분히 다르며 구분할 수 있음

Pred Prob를 보면 대부분 70을 넘어가는데, 이로부터 softmax probability를 confidence에 대한 direct representation으로 보면 안됨을 다시 한번 알 수 있음

#### 2.3.1. Sentiment Classification
![image](https://user-images.githubusercontent.com/49643709/160892678-3c7eb03b-5c16-4575-82b9-2e94cd63473f.png)
![image](https://user-images.githubusercontent.com/49643709/160893019-bd7bb4e1-069e-4881-bc4d-f7d9356ed51c.png)

Correct / incorrect, in-distribution / out-of-distribution을 구분할 수 있음

#### 2.3.2. Text Categorization
![image](https://user-images.githubusercontent.com/49643709/160893172-f3ce50fa-d45b-4f36-9781-c63a09d40528.png)
![image](https://user-images.githubusercontent.com/49643709/160893228-5e7d0c73-3e09-4b01-ba20-493c00baa970.png)

Correct / incorrect, in-distribution / out-of-distribution을 구분할 수 있음

#### 2.3.3. Part-of-Speech Tagging
![image](https://user-images.githubusercontent.com/49643709/160893252-8306bc64-9803-428c-8212-7b26aa63dfe4.png)
![image](https://user-images.githubusercontent.com/49643709/160893388-0f69aae5-be0e-42b2-b535-7d0678c7fb55.png)

Correct / incorrect, in-distribution / out-of-distribution을 구분할 수 있음

다만 weblog data는 tweet data보다 newswire style에 더 가깝기에, weblog sentence가 out-of-distribution인지 파악하는 것은 tweet보다 어려움

### 2.4. Automatic Speech Recognition
![image](https://user-images.githubusercontent.com/49643709/160893470-b0f9b228-625b-40ee-a839-f4518a214a00.png)

In-distribution / out-of-distribution을 구분할 수 있음

## 3. Abnormality Detection with Auxiliary Decoders
![image](https://user-images.githubusercontent.com/49643709/160902430-d9d8ca9e-2ced-4e91-9876-a0ecb53bc1cc.png)

1. In-distribution examples를 이용하여 blue layers 학습 (cross-entropy loss + reconstruction loss)
2. Freeze blue layers
3. Clean / noised examples를 이용하여 red layers 학습 (clean : normal, noised : abnormal)

Abnormality module (red layers)을 통해 baseline보다 더 좋은 성능을 얻을 수 있음

### 3.1. TIMIT
![image](https://user-images.githubusercontent.com/49643709/160903986-a10b9fdd-7da4-49c8-a3dc-991f986cdbbe.png)

Abnormality module이 학습하지 않은 noise에 대해서 test를 했음에도 효과적이었음

### 3.2. MNIST
![image](https://user-images.githubusercontent.com/49643709/160904698-e4a6499a-89a3-4e32-b4c1-f939c657a2a5.png)

Out-of-sample detection을 하는데 있어 softmax prediction probability보다 더 효과적임
